acc/components/guide.py

```diff
@@ -0,0 +1,293 @@
+                if pos_dur_next_wrapped:
+                    (pos_next_wrapped, dur_next_wrapped) = pos_dur_next_wrapped
+                    (pos_next, dur_next) = (pos_next_wrapped[0], dur_next_wrapped[0])
+                    if dur_next > 0 and dur_next < dur_min:
+                        ind_min = index
+                        (pos_min, dur_min) = (pos_next, dur_next)
+            if dur_min == math.inf:
+                # No new side was hit.
+                return None
+            duration += dur_min
+            if ind_min == 0:
+                # The first side is the exit from the guide.
+                return (pos_min, vel_curr, duration)
+            # Update particle to be reflecting from the new side.
+            pos_curr = pos_min
+            vel_curr = self.sides[ind_min].reflect(vel_curr)
+            ind_prev = ind_min
+
+    def calc_reflectivity(self, velocity_i, velocity_f):
+        """
+        Calculates the mirror reflectivity
+
+        Parameters:
+        velocity_i: initial velocities before reflecting
+        velocity_f: final velocities after reflecting
+
+        Returns
+        Reflectivity for each neutron
+        """
+        # see eq 5.2 in mcstas manual, sec 5.1.1
+        Q = self.m_over_h * numpy.linalg.norm(
+            velocity_i - velocity_f, axis=1)  # length of scattering vector
+        R = numpy.where(Q > self.Qc, 0.5 * self.R0 * (
+                1.0 - numpy.tanh((Q - self.m * self.Qc) / self.W) * (
+                    1.0 - self.alpha * (Q - self.Qc))),
+                        self.R0)
+        return R
+
+    # To do: use propagate method to implement process method
+    def process(self, neutrons):
+        if not len(neutrons):
+            return
+        arr = neutrons_as_npyarr(neutrons)
+        arr.shape = -1, ndblsperneutron
+
+        position = arr[:, 0:3]  # x, y, z
+        velocity = arr[:, 3:6]  # vx, vy, vz
+        time = arr[:, 8].reshape((arr.shape[0], 1))
+        prob = arr[:, 9].reshape((arr.shape[0], 1))
+
+        # initialize arrays containing neutron duration and side index
+        side = numpy.full((arr.shape[0], 1), -2, dtype=int)
+        old_side = side.copy()  # might not be necessary?
+        new_duration = numpy.full((arr.shape[0], 1), numpy.inf)
+
+        # Iterate until all neutrons hit end of guide or are absorbed
+        iter = 0
+        while numpy.count_nonzero(side == 0) + numpy.count_nonzero(
+                side == -1) != len(neutrons):
+            duration = numpy.full((arr.shape[0], 1), numpy.inf)
+            # Calculate minimum intersection time with all sides of the guide
+            for s in range(0, len(self.sides)):
+                intersection = self.sides[s].intersection_duration(position, velocity)
+                new_duration = numpy.minimum(intersection[1], duration,
+                                             where=((intersection[1] > 1e-10) & (s != old_side) & (old_side != 0)),
```

Comment:
@ckendrick: I believe that we have a bug here. My previous `propagate` was written when intersections occurring in the past were not returned at all. The guide planes are unbounded. For particles that do not even enter the guide, I think you're getting intersections with the planes before the guides even "start". There should probably be another condition here that the *z* of the intersection is positive.

Reply:
@mtbc I didn't realize the guide planes were unbounded. I agree that this is probably happening. Adding a z > 0 condition might not catch all the cases when the guide entrance is missed. I am thinking we need to define the entrance plane and initially test whether there is an intersection within the entrance bounds. If there isn't, then the particle will get removed before the while loop starts.

Reply:
@ckendrick: Good point, we may have to be far more defensive against inputs. @yxqd: Must we properly handle cases such as a source already inside the guide, or beside the guide, or suchlike? A strongly angled guide could perhaps have a neutron enter than bounce it back out of the entrance. Perhaps a general approach might be something like:

1. Test which neutrons are already within the guide. For those within, jump to step (3).
2. Neutrons outside the guide: Discard any neutron whose *first* intersection is with one of the five other planes rather than with the *bounded* entrance plane. The remaining neutrons, which do enter the guide, have them now start out from the entrance.
3. The iterative step: Include the unbounded entrance plane among the bouncing-around-inside. Discard any neutron hitting it as its next intersection.

This raises the question of what MCViNE does now.

Reply:
Looking back at the kickoff slides, the current fix may suffice for the above concerns.